consensus/polybft/stake_manager.go

```diff
@@ -0,0 +1,368 @@
+			}
+
+			var transferEvent contractsapi.TransferEvent
+
+			doesMatch, err := transferEvent.ParseLog(convertLog(log))
+			if err != nil {
+				return nil, err
+			}
+
+			if !doesMatch {
+				continue
+			}
+
+			events = append(events, &transferEvent)
+		}
+	}
+
+	return events, nil
+}
+
+// getValidatorInfo returns data for new validator (bls key, is active) from the supernet contract
+func (s *stakeManager) getNewValidatorInfo(address types.Address, stake *big.Int) (*ValidatorMetadata, error) {
+	encoded, err := getValidatorABI.Encode([]interface{}{address})
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := s.rootChainRelayer.Call(
+		s.key.Address(),
+		ethgo.Address(s.supernetManagerContract),
+		encoded)
+	if err != nil {
+		return nil, fmt.Errorf("failed to invoke validators function on the supernet manager: %w", err)
+	}
+
+	byteResponse, err := hex.DecodeHex(response)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode hex response, %w", err)
+	}
+
+	decoded, err := getValidatorABI.Outputs.Decode(byteResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	//nolint:godox
+	// TODO - @goran-ethernal change this to use the generated stub
+	// once we remove old ChildValidatorSet stubs and generate new ones
+	// from the new contract
+	output, ok := decoded.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("could not convert decoded outputs to map")
+	}
+
+	outputMap, ok := output["0"].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("could not convert decoded outputs to map")
+	}
+
+	blsKey, ok := outputMap["blsKey"].([4]*big.Int)
+	if !ok {
+		return nil, fmt.Errorf("failed to decode blskey")
+	}
+
+	pubKey, err := bls.UnmarshalPublicKeyFromBigInt(blsKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal BLS public key: %w", err)
+	}
+
+	return &ValidatorMetadata{
+		Address:     address,
+		VotingPower: stake,
+		BlsKey:      pubKey,
+		IsActive:    true,
+	}, nil
+}
+
+// stakeInfo holds info about validator stake
+// it holds sorted map of all stakes of all validators
+type stakeInfo struct {
+	pos     int
+	stake   *big.Int
+	address types.Address
+}
+
+// stakeCOunter sorts and returns stake info for all validators
+type stakeCounter struct {
+	stakeMap            map[types.Address]*stakeInfo
+	currentValidatorSet AccountSet
+}
+
+// newStakeCounter returns a new instance of stake counter
+func newStakeCounter(currentValidatorSet AccountSet) *stakeCounter {
+	stakeCounter := &stakeCounter{
+		currentValidatorSet: currentValidatorSet,
+		stakeMap:            make(map[types.Address]*stakeInfo, 0),
+	}
+
+	for _, v := range currentValidatorSet {
+		stakeCounter.stakeMap[v.Address] = &stakeInfo{
+			stake:   new(big.Int).Set(v.VotingPower),
+			address: v.Address,
+		}
+	}
+
+	return stakeCounter
+}
+
+// addStake adds given amount for a validator to stake map
+func (sc *stakeCounter) addStake(address types.Address, amount *big.Int) {
+	sInfo, exists := sc.stakeMap[address]
```

Comment:
Maybe it would be even less code:
```
	if sInfo, exists := sc.stakeMap[address]; exists {
    	        sInfo.stake.Add(sInfo.stake, amount)
	} else {
		sc.stakeMap[address] = &stakeInfo{address: address, stake: amount}
	}

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/commit/ce5d87c88920b208f4834629f6fe9c135d997e4f